src/test/txrequest_tests.cpp

```diff
@@ -0,0 +1,587 @@
+    // high priority to low priority within each class.
+    for (int i = peers - 1; i >= 0; --i) {
+        scenario.AdvanceTime(reqtimes[request_order[i]] - scenario.Now() - MICROSECOND);
+        scenario.Check(request_order[i], {}, 1, 0, 0, "b2");
+        scenario.AdvanceTime(MICROSECOND);
+        scenario.Check(request_order[i], {gtxid}, 1, 0, 0, "b3");
+    }
+
+    // Peers now in random order go offline, or send NOTFOUNDs. At every point in time the new to-be-requested-from
+    // peer should be the best remaining one, so verify this after every response.
+    for (int i = 0; i < peers; ++i) {
+        if (InsecureRandBool()) scenario.AdvanceTime(RandomTime8s());
+        const int pos = InsecureRandRange(request_order.size());
+        const auto peer = request_order[pos];
+        request_order.erase(request_order.begin() + pos);
+        if (InsecureRandBool()) {
+            scenario.DisconnectedPeer(peer);
+            scenario.Check(peer, {}, 0, 0, 0, "b4");
+        } else {
+            scenario.ReceivedResponse(peer, gtxid.GetHash());
+            scenario.Check(peer, {}, 0, 0, request_order.size() > 0, "b5");
+        }
+        if (request_order.size()) {
+            scenario.Check(request_order[0], {gtxid}, 1, 0, 0, "b6");
+        }
+    }
+
+    // Everything is gone in the end.
+    for (const auto peer : announce_order) {
+        scenario.Check(peer, {}, 0, 0, 0, "b7");
+    }
+}
+
+/** Add to scenario a test with one peer announcing two transactions, to verify they are
+ *  fetched in announcement order. */
+void BuildRequestOrderTest(Scenario& scenario, int config)
+{
+    scenario.SetTestName(strprintf("RequestOrder(config=%i)", config));
+
+    auto peer = scenario.NewPeer();
+    auto gtxid1 = scenario.NewGTxid();
+    auto gtxid2 = scenario.NewGTxid();
+
+    auto reqtime2 = scenario.Now() + RandomTime8s();
+    auto reqtime1 = reqtime2 + RandomTime8s();
+
+    scenario.ReceivedInv(peer, gtxid1, config & 1, reqtime1);
+    // Simulate time going backwards by giving the second announcement an earlier reqtime.
+    scenario.ReceivedInv(peer, gtxid2, config & 2, reqtime2);
+
+    scenario.AdvanceTime(reqtime2 - MICROSECOND - scenario.Now());
+    scenario.Check(peer, {}, 2, 0, 0, "o1");
+    scenario.AdvanceTime(MICROSECOND);
+    scenario.Check(peer, {gtxid2}, 2, 0, 0, "o2");
+    scenario.AdvanceTime(reqtime1 - MICROSECOND - scenario.Now());
+    scenario.Check(peer, {gtxid2}, 2, 0, 0, "o3");
+    scenario.AdvanceTime(MICROSECOND);
+    // Even with time going backwards in between announcements, the return value of GetRequestable is in
+    // announcement order.
+    scenario.Check(peer, {gtxid1, gtxid2}, 2, 0, 0, "o4");
+
+    scenario.DisconnectedPeer(peer);
+    scenario.Check(peer, {}, 0, 0, 0, "o5");
+}
+
+/** Add to scenario a test thats verifies behavior related to both txid and wtxid with the same
```

Comment:
ea7839c8889864be174b85a1f4de326f5e12e430, typo:
```suggestion
/** Add to scenario a test that verifies behavior related to both txid and wtxid with the same
```

Reply:
Done.